dev/plugins/hu.elte.txtuml.api.model.execution/src/hu/elte/txtuml/api/model/execution/impl/seqdiag/Message.java

```diff
@@ -4,80 +4,86 @@
 
 @SequenceDiagramRelated
-public class Message extends WrapperImpl<Signal> {
-	private final ModelClass target;
-	private final ModelClass sender;
-	private final Optional<Boolean> fromActor;
-
-	public static Message fromObject(ModelClass sender, Signal signal, ModelClass target) {
-		return new Message(sender, signal, target, false);
-	}
-
-	public static Message fromActor(Signal signal, ModelClass target) {
-		return new Message(null, signal, target, true);
-	}
+public class Message<T extends ModelClass, U extends ModelClass> extends AbstractMessage {
 
-	public static Message fromUnknown(Signal signal, ModelClass target) {
-		return new Message(null, signal, target, null);
-	}
+	private final MessageParticipant<T> sender;
+	private final MessageParticipant<U> target;
+	private final Optional<Boolean> fromActor;
 
-	protected Message(ModelClass sender, Signal signal, ModelClass target, Boolean fromActor) {
+	protected Message(MessageParticipant<T> sender, Signal signal, MessageParticipant<U> target, Boolean fromActor) {
 		super(signal);
 		this.target = target;
 		this.sender = sender;
 		this.fromActor = Optional.ofNullable(fromActor);
 	}
 
-	public ModelClass getTarget() {
+	public MessageParticipant<U> getTarget() {
 		return target;
 	}
 
-	public ModelClass getSenderOrNull() {
+	public MessageParticipant<T> getSenderOrNull() {
 		return sender;
 	}
 
 	public Optional<Boolean> isFromActor() {
 		return fromActor;
 	}
 
-	public boolean matches(Message other) {
-		if (other == this) {
+	public <T2 extends ModelClass, U2 extends ModelClass> boolean matches(Message<T2, U2> expected) {
+		// this: actual, other: expected
+		if (expected == this) {
 			return true;
 		}
-		if (other == null) {
+		if (expected == null) {
 			return false;
 		}
 
-		if (fromActor.isPresent() && other.fromActor.isPresent()) {
-			if (!fromActor.get().equals(other.fromActor.get())) {
-				return false;
-			}
-			if (sender != other.sender) { // both is null if from actor
-				return false;
-			}
+		if (fromActor.isPresent() && !fromActor.equals(expected.fromActor)) {
+			return false;
 		}
-		if (target != other.target) {
+
+		if (sender != null && expected.sender != null && expected.sender.hasParticipant()
+				&& sender.getParticipant().get() != expected.sender.getParticipant().get()) {
 			return false;
 		}
 
+		if (target != null && expected.target != null && expected.target.hasParticipant()
+				&& target.getParticipant().get() != expected.target.getParticipant().get()) {
+			return false;
+		}
+
+		boolean areSignalsEqual;
 		try {
-			return BasedOnFields.equal(getWrapped(), other.getWrapped());
+			areSignalsEqual = BasedOnFields.equal(getWrapped(), expected.getWrapped());
 		} catch (IllegalAccessException e) {
 			return false;
 		}
+
+		if (areSignalsEqual) {
+			if (sender != null && expected.sender != null && !expected.sender.hasParticipant()) {
+				expected.sender.bindTo(sender.getParticipant().get());
```

Comment:
I think we should check the classes of the senders before binding, because they might not be the same and in this case we have an error.

Reply:
I agree, thank you!